core/src/main/java/org/springframework/security/oauth2/server/authorization/OAuth2Authorization.java

```diff
@@ -15,17 +15,127 @@
+		this(registeredClientId, principalName, null, attributes);
+	}
+
+	/**
+	 * Creates an {@link OAuth2Authorization} object with the provided params and an empty, unmodifiable
+	 * {@code attributes} map.
+	 *
+	 * @see #OAuth2Authorization(String, String, OAuth2AccessToken, Map)
+	 */
+	public OAuth2Authorization(String registeredClientId, String principalName, OAuth2AccessToken accessToken) {
+		this(registeredClientId, principalName, accessToken, Collections.emptyMap());
+	}
+
+	/**
+	 * Creates an {@link OAuth2Authorization} object with the provided params.
+	 *
+	 * @param registeredClientId the client's identifier which issued the authorization
+	 * @param principalName      the name of the principal the client wants to authorize
+	 * @param accessToken        the {@link OAuth2AccessToken access token} provided for authorization
+	 * @param attributes         additional attributes associated with the authorization
+	 */
+	public OAuth2Authorization(String registeredClientId, String principalName, @Nullable OAuth2AccessToken accessToken,
+			@Nullable Map<String, Object> attributes) {
+		Assert.hasText(registeredClientId, "registeredClientId cannot be empty");
+		Assert.hasText(principalName, "principalName cannot be empty");
+		this.registeredClientId = registeredClientId;
+		this.principalName = principalName;
+		this.accessToken = accessToken;
+		this.attributes = attributes == null ?
+				Collections.emptyMap() :
+				Collections.unmodifiableMap(attributes);
+		this.tokenType = determineTokenType(this.accessToken, this.attributes);
+	}
+
+	public String getRegisteredClientId() {
+		return registeredClientId;
+	}
+
+	public String getPrincipalName() {
+		return principalName;
+	}
+
+	public OAuth2AccessToken getAccessToken() {
+		return accessToken;
+	}
+
+	public Map<String, Object> getAttributes() {
+		return attributes;
+	}
+
+	public TokenType getTokenType() {
+		return tokenType;
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+		OAuth2Authorization that = (OAuth2Authorization) o;
+		return Objects.equals(registeredClientId, that.registeredClientId) &&
+				Objects.equals(principalName, that.principalName) &&
+				Objects.equals(accessToken, that.accessToken) &&
+				Objects.equals(attributes, that.attributes) &&
+				Objects.equals(tokenType, that.tokenType);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(registeredClientId, principalName, accessToken, attributes, tokenType);
+	}
+
+	@Override
+	public String toString() {
+		return "OAuth2Authorization{" +
+				"registeredClientId='" + registeredClientId + '\'' +
+				", principalName='" + principalName + '\'' +
+				", accessToken=" + accessToken +
+				", attributes=" + attributes +
+				", tokenType=" + tokenType +
+				'}';
+	}
 
+	private TokenType determineTokenType(OAuth2AccessToken accessToken, Map<String, Object> attributes) {
```

Comment:
If I understand correctly, there is either an `OAuth2AccessToken` set, or an `authorization_code` entry should exist in the `attributes` map. In this case I assume it would be best to verify it during instantiation, hence the solution. Should this be done elsewhere, e.g. a dedicated class? Either way, this feels really clunky.
Another solution I can think of would be to create an interface `OAuth2Authorization` which has methods `getToken()` and `getTokenType()` with 2 implementations, `AccessTokenOAuth2Authorization` and `AuthorizationCodeOAuth2Authorization`, or something along these lines. Thoughts?

Reply:
From my opinion creating the interface is definetely the better solution. Besides the usage of bearer tokens (RFC6750) for example is simply a recommendation of RFC6749. The structure of an access token is not defined at all and it is allowed to use any type of token. I do not know of any case in which someone used another token type as bearer but if someone would need an alternative to the bearer token another approach than the interface would be problematic.

Reply:
Thanks for the input. I updated the PR with what I think is a cleaner solution.